File: packages/python/src/i18n/lang.py
from __future__ import annotations

import json
import os
import re
from pathlib import Path
from typing import Any, Literal, TypedDict, cast

SupportedLang = Literal["en-US", "zh-CN"]

_LOCALE_DIR = Path(__file__).resolve().parent.parent / "locale"


def _load_bundle(lang: SupportedLang) -> dict[str, Any]:
    path = _LOCALE_DIR / ("zh-CN.json" if lang == "zh-CN" else "en-US.json")
    with path.open(encoding="utf-8") as f:
        return cast(dict[str, Any], json.load(f))


def normalize_lang(raw: Any) -> SupportedLang:
    if raw == "zh-CN":
        return "zh-CN"
    if raw == "en-US":
        return "en-US"
    return "en-US"


def supported_lang_from_lang_env(raw: str | None) -> SupportedLang:
    if raw is None:
        return "en-US"
    s = raw.strip()
    if not s:
        return "en-US"
    lower = s.lower()
    if lower in ("c", "posix"):
        return "en-US"
    if re.match(r"^zh([_.-]|$)", s, re.I):
        return "zh-CN"
    if re.match(r"^en([_.-]|$)", s, re.I):
        return "en-US"
    return "en-US"


def get_tool_metadata_locale() -> SupportedLang:
    return supported_lang_from_lang_env(os.environ.get("LANG"))


def port_lang() -> SupportedLang:
    return supported_lang_from_lang_env(os.environ.get("LANG"))


def interpolate(template: str, vars: dict[str, str | int] | None = None) -> str:
    if not vars:
        return template
    s = template
    for k, v in vars.items():
        s = str(v).join(s.split(f"{{{{{k}}}}}"))
    return s


class _ServerFlat(TypedDict, total=False):
    browser_not_installed_message: str
    browser_not_installed_hint: str
    extension_not_connected_message: str
    extension_not_connected_hint: str
    extension_not_connected_troubleshoot_1: str
    extension_not_connected_troubleshoot_2: str
    extension_not_connected_troubleshoot_3: str
    extension_not_connected_troubleshoot_4: str
    extension_not_connected_troubleshoot_5: str
    server_draining_requests: str
    server_draining_new_tasks: str
    bridge_timeout: str
    disconnect_replaced: str
    disconnect_extension_gone: str
    disconnect_server_shutdown: str
    mcp_tool_response_truncated_after_limit: str
    daemon_unreachable_message: str
    daemon_unreachable_hint: str


def _flatten_server_messages(sm: dict[str, Any]) -> dict[str, str]:
    enc = sm["extensionNotConnected"]["troubleshoot"]
    return {
        "browser_not_installed.message": sm["browserNotInstalled"]["message"],
        "browser_not_installed.hint": sm["browserNotInstalled"]["hint"],
        "extension_not_connected.message": sm["extensionNotConnected"]["message"],
        "extension_not_connected.hint": sm["extensionNotConnected"]["hint"],
        "extension_not_connected.troubleshoot.1": enc["1"],
        "extension_not_connected.troubleshoot.2": enc["2"],
        "extension_not_connected.troubleshoot.3": enc["3"],
        "extension_not_connected.troubleshoot.4": enc["4"],
        "extension_not_connected.troubleshoot.5": enc["5"],
        "server_draining.requests": sm["serverDraining"]["requests"],
        "server_draining.new_tasks": sm["serverDraining"]["new_tasks"],
        "bridge_timeout": sm["bridge"]["timeout"],
        "disconnect.replaced": sm["disconnect"]["replaced"],
        "disconnect.extension_gone": sm["disconnect"]["extension_gone"],
        "disconnect.server_shutdown": sm["disconnect"]["server_shutdown"],
        "mcp_tool.response_truncated_after_limit": sm["mcpTool"]["responseTruncatedAfterLimit"],
        "daemon_unreachable.message": sm["daemonUnreachable"]["message"],
        "daemon_unreachable.hint": sm["daemonUnreachable"]["hint"],
    }


MessageId = str
_by_lang: dict[SupportedLang, dict[str, str]] = {
    "en-US": _flatten_server_messages(_load_bundle("en-US")["serverMessages"]),
    "zh-CN": _flatten_server_messages(_load_bundle("zh-CN")["serverMessages"]),
}

_bundles: dict[SupportedLang, dict[str, Any]] = {
    "en-US": _load_bundle("en-US"),
    "zh-CN": _load_bundle("zh-CN"),
}


def t(lang: SupportedLang, id: str, vars: dict[str, str | int] | None = None) -> str:
    table = _by_lang.get(lang) or _by_lang["en-US"]
    template = table.get(id) or _by_lang["en-US"].get(id) or id
    return interpolate(template, vars)


PortMessageKey = str


def port_t(lang: SupportedLang, key: str, vars: dict[str, str | int] | None = None) -> str:
    bundle = _bundles.get(lang) or _bundles["en-US"]
    template = bundle["port"][key]
    return interpolate(template, vars)


BridgeProtocolKey = str


def bridge_t(lang: SupportedLang, key: str, vars: dict[str, str | int] | None = None) -> str:
    bundle = _bundles.get(lang) or _bundles["en-US"]
    row = bundle["bridgeProtocol"]
    template = row[key]
    return interpolate(template, vars)


LogMessageKey = str


def log_t(lang: SupportedLang, key: str, vars: dict[str, str | int] | None = None) -> str:
    bundle = _bundles.get(lang) or _bundles["en-US"]
    template = bundle["logMessages"][key]
    return interpolate(template, vars)


def tools_bundle(locale: SupportedLang) -> dict[str, Any]:
    b = _bundles.get(locale) or _bundles["en-US"]
    return cast(dict[str, Any], b["tools"])


def mcp_context_bundle(locale: SupportedLang) -> dict[str, Any]:
    b = _bundles.get(locale) or _bundles["en-US"]
    return cast(dict[str, Any], b["mcpContext"])

File: packages/node/src/i18n/lang.ts
import enUS from '../locale/en-US.json' with { type: 'json' };
import zhCN from '../locale/zh-CN.json' with { type: 'json' };

export type SupportedLang = 'en-US' | 'zh-CN';

/** BCP 47 tags supported by MCP Server; anything else falls back to en-US. */
export function normalizeLang(raw: unknown): SupportedLang {
  if (raw === 'zh-CN') return 'zh-CN';
  if (raw === 'en-US') return 'en-US';
  return 'en-US';
}

/**
 * Maps process **LANG** (MCP `env` or OS) to tool-list metadata language.
 * Accepts BCP 47 (`zh-CN`) and common POSIX forms (`zh_CN.UTF-8`, `en_US.UTF-8`).
 */
export function supportedLangFromLangEnv(raw: string | undefined): SupportedLang {
  if (raw == null) return 'en-US';
  const s = raw.trim();
  if (!s) return 'en-US';
  const lower = s.toLowerCase();
  if (lower === 'c' || lower === 'posix') return 'en-US';
  if (/^zh([_.-]|$)/i.test(s)) return 'zh-CN';
  if (/^en([_.-]|$)/i.test(s)) return 'en-US';
  return 'en-US';
}

/**
 * Locale for MCP **tool list** metadata (`description` + Zod `describe` strings).
 * Set in MCP Client config, e.g. `"env": { "LANG": "zh-CN" }` (or OS `LANG` like `zh_CN.UTF-8`).
 */
export function getToolMetadataLocale(): SupportedLang {
  return supportedLangFromLangEnv(process.env.LANG);
}

export type MessageId =
  | 'browser_not_installed.message'
  | 'browser_not_installed.hint'
  | 'extension_not_connected.message'
  | 'extension_not_connected.hint'
  | 'extension_not_connected.troubleshoot.1'
  | 'extension_not_connected.troubleshoot.2'
  | 'extension_not_connected.troubleshoot.3'
  | 'extension_not_connected.troubleshoot.4'
  | 'extension_not_connected.troubleshoot.5'
  | 'extension_not_connected.store_opened_hint'
  | 'server_draining.requests'
  | 'server_draining.new_tasks'
  | 'bridge_timeout'
  | 'disconnect.replaced'
  | 'disconnect.extension_gone'
  | 'disconnect.server_shutdown'
  | 'mcp_tool.response_truncated_after_limit'
  | 'http_fetch_fallback.note'
  | 'daemon_unreachable.message'
  | 'daemon_unreachable.hint';

export type PortMessageKey = keyof typeof enUS.port;

export type BridgeProtocolKey = keyof typeof enUS.bridgeProtocol;

export type LogMessageKey = keyof typeof enUS.logMessages;

type ServerMessages = typeof enUS.serverMessages;

function flattenServerMessages(sm: ServerMessages): Record<MessageId, string> {
  return {
    'browser_not_installed.message': sm.browserNotInstalled.message,
    'browser_not_installed.hint': sm.browserNotInstalled.hint,
    'extension_not_connected.message': sm.extensionNotConnected.message,
    'extension_not_connected.hint': sm.extensionNotConnected.hint,
    'extension_not_connected.troubleshoot.1': sm.extensionNotConnected.troubleshoot['1'],
    'extension_not_connected.troubleshoot.2': sm.extensionNotConnected.troubleshoot['2'],
    'extension_not_connected.troubleshoot.3': sm.extensionNotConnected.troubleshoot['3'],
    'extension_not_connected.troubleshoot.4': sm.extensionNotConnected.troubleshoot['4'],
    'extension_not_connected.troubleshoot.5': sm.extensionNotConnected.troubleshoot['5'],
    'extension_not_connected.store_opened_hint': sm.extensionNotConnected.storeOpenedHint,
    'server_draining.requests': sm.serverDraining.requests,
    'server_draining.new_tasks': sm.serverDraining.new_tasks,
    'bridge_timeout': sm.bridge.timeout,
    'disconnect.replaced': sm.disconnect.replaced,
    'disconnect.extension_gone': sm.disconnect.extension_gone,
    'disconnect.server_shutdown': sm.disconnect.server_shutdown,
    'mcp_tool.response_truncated_after_limit': sm.mcpTool.responseTruncatedAfterLimit,
    'http_fetch_fallback.note': sm.httpFetchFallback.note,
    'daemon_unreachable.message': sm.daemonUnreachable.message,
    'daemon_unreachable.hint': sm.daemonUnreachable.hint,
  };
}

const byLang: Record<SupportedLang, Record<MessageId, string>> = {
  'en-US': flattenServerMessages(enUS.serverMessages),
  'zh-CN': flattenServerMessages(zhCN.serverMessages),
};

export function interpolate(template: string, vars?: Record<string, string | number>): string {
  if (!vars) return template;
  let s = template;
  for (const [k, v] of Object.entries(vars)) {
    s = s.split(`{{${k}}}`).join(String(v));
  }
  return s;
}

/** Resolve a server-side user-visible string for the given BCP 47 tag (see docs/mcp.md §5.0). */
export function t(lang: SupportedLang, id: MessageId, vars?: Record<string, string | number>): string {
  const table = byLang[lang] ?? byLang['en-US'];
  const template = table[id] ?? byLang['en-US'][id];
  return interpolate(template, vars);
}

const bundles: Record<SupportedLang, typeof enUS> = {
  'en-US': enUS,
  'zh-CN': zhCN,
};

/** Operator-facing port / startup messages (follows `LANG` like tool metadata). */
export function portT(
  lang: SupportedLang,
  key: PortMessageKey,
  vars?: Record<string, string | number>,
): string {
  const bundle = bundles[lang] ?? bundles['en-US'];
  return interpolate(bundle.port[key], vars);
}

export function portLang(): SupportedLang {
  return supportedLangFromLangEnv(process.env.LANG);
}

const bridgeByLang: Record<SupportedLang, Record<BridgeProtocolKey, string>> = {
  'en-US': enUS.bridgeProtocol as Record<BridgeProtocolKey, string>,
  'zh-CN': zhCN.bridgeProtocol as Record<BridgeProtocolKey, string>,
};

/** WebSocket bridge JSON-RPC / close-reason strings (follows `LANG`; no per-tool `lang`). */
export function bridgeT(
  lang: SupportedLang,
  key: BridgeProtocolKey,
  vars?: Record<string, string | number>,
): string {
  const row = bridgeByLang[lang] ?? bridgeByLang['en-US'];
  const template = row[key] ?? bridgeByLang['en-US'][key];
  return interpolate(template, vars);
}

const logByLang: Record<SupportedLang, Record<LogMessageKey, string>> = {
  'en-US': enUS.logMessages as Record<LogMessageKey, string>,
  'zh-CN': zhCN.logMessages as Record<LogMessageKey, string>,
};

/** stderr log lines (follows `LANG`). */
export function logT(
  lang: SupportedLang,
  key: LogMessageKey,
  vars?: Record<string, string | number>,
): string {
  const row = logByLang[lang] ?? logByLang['en-US'];
  const template = row[key] ?? logByLang['en-US'][key];
  return interpolate(template, vars);
}
